pyteal/ast/abi/transaction.py

```diff
@@ -0,0 +1,196 @@
+from abc import abstractmethod
+from typing import TypeVar, Union, cast
+from pyteal.ast.abi.type import BaseType, ComputedValue, TypeSpec
+from pyteal.ast.expr import Expr
+from pyteal.ast.int import Int
+from pyteal.ast.txn import TxnType
+from pyteal.types import TealType
+from pyteal.errors import TealInputError
+
+T = TypeVar("T", bound=BaseType)
+
+
+class TransactionTypeSpec(TypeSpec):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def new_instance(self) -> "Transaction":
+        return Transaction(self)
+
+    def annotation_type(self) -> "type[Transaction]":
+        return Transaction
+
+    def is_dynamic(self) -> bool:
+        return False
+
+    def byte_length_static(self) -> int:
+        raise TealInputError("Transaction Types don't have a static size")
+
+    def storage_type(self) -> TealType:
+        return TealType.uint64
+
+    def __eq__(self, other: object) -> bool:
+        return type(self) is type(other)
+
+    @abstractmethod
+    def __str__(self) -> str:
+        return TxnType.Any.name
+
+
+TransactionTypeSpec.__module__ = "pyteal"
+
+
+class Transaction(BaseType):
+    def __init__(self, spec: TransactionTypeSpec) -> None:
+        super().__init__(spec)
+
+    def type_spec(self) -> TransactionTypeSpec:
+        return cast(TransactionTypeSpec, super().type_spec())
+
+    def get(self) -> Expr:
+        return self.stored_value.load()
+
+    def set(self: T, value: Union[int, Expr, "Transaction", ComputedValue[T]]) -> Expr:
+        match value:
+            case ComputedValue():
+                self._set_with_computed_type(value)
+            case BaseType():
+                return self.stored_value.store(self.stored_value.load())
+            case int():
+                return self.stored_value.store(Int(value))
+            case Expr():
+                return self.stored_value.store(value)
+            case _:
+                raise TealInputError(f"Cant store a {type(value)} in a Transaction")
+
+    def validate(self) -> Expr:
+        # TODO: make sure the group length is large enough and the index is valid?
+        pass
+
+    def decode(
+        self,
+        encoded: Expr,
+        *,
+        startIndex: Expr = None,
+        endIndex: Expr = None,
+        length: Expr = None,
+    ) -> Expr:
+        raise TealInputError("A Transaction cannot be decoded")
+
+    def encode(self) -> Expr:
+        raise TealInputError("A Transaction cannot be encoded")
+
+
+Transaction.__module__ = "pyteal"
+
+
+class AnyTransactionTypeSpec(TransactionTypeSpec):
```

Comment:
Why have `Transaction` and `AnyTransaction`? I think it'd make more sense from an inheritance perspective to get rid of `AnyTransaction` and have `Transaction` be used when it's needed. That way a `PaymentTransaction` and is a `Transaction`.

Reply:
you're right, I think I just tripped over some python typing thing and punted. Will make this work.